File: apple/QuenderinKit/Sources/QuenderinKit/ScriptedInferenceEngine.swift
import Foundation

/// An `InferenceEngine` that returns a fixed queue of replies in order — so
/// multi-turn agent loops can be tested (and previewed) deterministically,
/// without a real model. Once the script is exhausted it returns a final-answer
/// JSON so loops terminate cleanly.
public actor ScriptedInferenceEngine: InferenceEngine {
    private var replies: [String]
    private var loaded: String? = "scripted"

    public init(replies: [String]) {
        self.replies = replies
    }

    public func loadedModelID() async -> String? { loaded }
    public func load(model: ModelEntry, at fileURL: URL) async throws { loaded = model.id }
    public func unload() async { loaded = nil }

    public func generate(prompt: String, options: GenerationOptions) async throws -> AsyncThrowingStream<String, Error> {
        let next = replies.isEmpty ? "{\"answer\":\"(no more scripted replies)\"}" : replies.removeFirst()
        return AsyncThrowingStream { continuation in
            continuation.yield(next)
            continuation.finish()
        }
    }
}

File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/ScriptedInferenceEngine.kt
package ai.quenderin.core

/**
 * Deterministic engine that replays scripted replies in order — for testing the agent loop
 * (and multi-turn flows) without real inference. Mirrors iOS `ScriptedInferenceEngine`.
 */
class ScriptedInferenceEngine(private val replies: List<String>) : InferenceEngine {
    private var index = 0

    override var loadedModelId: String? = "scripted"
        private set

    override fun load(model: ModelEntry, filePath: String) {
        loadedModelId = model.id
    }

    override fun unload() {
        loadedModelId = null
    }

    override fun complete(prompt: String): String {
        val reply = replies.getOrNull(index) ?: """{"answer":"(no more scripted replies)"}"""
        index++
        return reply
    }
}
